contrib/gorilla/websocket/websocket.go

```diff
@@ -0,0 +1,143 @@
+	span, ctx := tracer.StartSpanFromContext(r.Context(), "websocket.connection")
+	defer func() {
+		// Finish the span when an error occurred.
+		// Otherwise, finish it in the (*Conn).Close() method.
+		if err != nil {
+			span.Finish(tracer.WithError(err))
+		}
+	}()
+
+	tracee, err := u.Upgrader.Upgrade(w, r.WithContext(ctx), responseHeader)
+	if err != nil {
+		return nil, err
+	}
+	return wrapConn(ctx, tracee), nil
+}
+
+// Conn is the wrapper type of a traced connection.
+type Conn struct {
+	*websocket.Conn
+	ctx context.Context
+}
+
+func wrapConn(ctx context.Context, conn *websocket.Conn) *Conn {
+	return &Conn{
+		Conn: conn,
+		ctx:  ctx,
+	}
+}
+
+// Close wraps the actual connection Close() method and finishes the
+// websocket.connection span started by Upgrade().
+func (c *Conn) Close() (err error) {
+	if span, ok := tracer.SpanFromContext(c.ctx); ok {
+		defer func() {
+			span.Finish(tracer.WithError(err))
+		}()
+	}
+	return c.Conn.Close()
+}
+
+// ReadMessage traces the actual ReadMessage() method using the global tracer
+// as span websocket.read_message.
+func (c *Conn) ReadMessage() (messageType int, p []byte, err error) {
+	span, _ := tracer.StartSpanFromContext(c.ctx, "websocket.read_message", messageTypeTag(messageType))
+	defer func() {
+		span.Finish(tracer.WithError(err))
+	}()
+	return c.Conn.ReadMessage()
+}
+
+// ReadJSON traces the actual ReadJSON() method using the global tracer
+// as span websocket.read_json.
+func (c *Conn) ReadJSON(v interface{}) (err error) {
+	span, _ := tracer.StartSpanFromContext(c.ctx, "websocket.read_json")
+	defer func() {
+		span.Finish(tracer.WithError(err))
+	}()
+	return c.Conn.ReadJSON(v)
+}
+
+// WriteMessage traces the actual WriteMessage() method using the global tracer
+// as span websocket.write_message.
+func (c *Conn) WriteMessage(messageType int, data []byte) (err error) {
+	span, _ := tracer.StartSpanFromContext(c.ctx, "websocket.write_message",
+		messageTypeTag(messageType), messageLengthTag(len(data)))
+	defer func() {
+		span.Finish(tracer.WithError(err))
+	}()
+	return c.Conn.WriteMessage(messageType, data)
+}
+
+// WritePreparedMessage traces the actual WritePreparedMessage() method using
+// the global tracer as span websocket.write_prepared_message.
+func (c *Conn) WritePreparedMessage(pm *websocket.PreparedMessage) (err error) {
+	span, _ := tracer.StartSpanFromContext(c.ctx, "websocket.write_prepared_message")
+	defer func() {
+		span.Finish(tracer.WithError(err))
+	}()
+	return c.Conn.WritePreparedMessage(pm)
+}
+
+// WriteControl traces the actual WriteControl() method using the global tracer
+// as span websocket.write_control.
+func (c *Conn) WriteControl(messageType int, data []byte, deadline time.Time) (err error) {
+	span, _ := tracer.StartSpanFromContext(c.ctx, "websocket.write_control",
+		messageTypeTag(messageType), messageLengthTag(len(data)))
```

Comment:
I don't really like this. Can we drop the helper methods? I don't see any benefit in them, other than having a reader search for their definition to understand what they do here. Without them, even if more verbose, the code can be understood on the first pass because it's common, known API, and not newly introduced concepts. Basically what I'm saying is:
```suggestion
	span, _ := tracer.StartSpanFromContext(c.ctx, 
		"websocket.write_control",
		tracer.Tag("websocket.message_type", mtype),
		tracer.Tag("websocket.message_length", len(data)),
	)
```
Same below. And in the spirit of having short variable names can we use `mtype` here? I don't see a reason for a long descriptive name. In the spirit of "[naming things](https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#naming-things)".

Reply:
Simple code factorization, in case the tag name had to be changed.

Reply:
I changed it and inlined the tracer.Tag calls, factoring out the tag names into global const string values.

Reply:
Perfect! THanks